### PR TITLE
add app mode testing server

### PR DESCRIPTION
**Pull request by ii2day**

### CODE REVIEW

**Comment by panpan0000** [comment]
1. can you elaborate the difference between this and k8s readinessProbe in the README. 
2. also doc the test summary of the distributed stress client.

**Comment by ii2day** [comment]
> 1. can you elaborate the difference between this and k8s readinessProbe in the README.
> 2. also doc the test summary of the distributed stress client.

1.This mode is used for kdoctor e2e testing, and readinessProbe still uses http
2.The report shows the request status of each client

**Comment by weizhoublue** [comment]
conflict

**Comment by ii2day** [comment]
> conflict

done

File: charts/templates/service.yaml [hunk 1]
```diff
@@ -29,7 +29,11 @@ spec:
     {{- end }}
     {{- if .Values.feature.enableIPv4 }}
     - name: http
-      port: {{ .Values.kdoctorAgent.httpServer.port }}
+      port: {{ .Values.kdoctorAgent.httpServer.healthPort }}
```
Comment (weizhoublue) [inline charts/templates/service.yaml:32]: health 不需要 用 service 暴露出来

File: cmd/agent/cmd/appHttpServer.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	"github.com/kdoctor-io/kdoctor/pkg/types"
+	"go.uber.org/zap"
+	"os"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	ParamInformation = map[string]string{
+		"delay": "in query, delay some second return response",
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:24]: delay  support all method ？

File: cmd/agent/cmd/appHttpServer.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+	requestCounts int64 = 0
+)
+
+// ---------- get
+type echoGetHandler struct {
+	logger *zap.Logger
+}
+
+func (s *echoGetHandler) Handle(r echo.GetParams) middleware.Responder {
+	s.logger.Sugar().Debugf("%s method  %s request from %s", r.HTTPRequest.Proto, r.HTTPRequest.Method, r.HTTPRequest.RemoteAddr)
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:35]: also print  r.Delay

File: cmd/agent/cmd/appHttpServer.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+		srvApp.TLSPort = int(types.AgentConfig.AgentAppPort)
+		if len(types.AgentConfig.TlsCaCertPath) != 0 {
+			srvApp.TLSCACertificate = flags.Filename(types.AgentConfig.TlsCaCertPath)
+		}
+		if len(types.AgentConfig.TlsServerCertPath) != 0 && len(types.AgentConfig.TlsServerKeyPath) != 0 {
+			srvApp.TLSCertificate = flags.Filename(types.AgentConfig.TlsServerCertPath)
+			srvApp.TLSCertificateKey = flags.Filename(types.AgentConfig.TlsServerKeyPath)
+		}
+	} else {
+		srvApp.Port = int(types.AgentConfig.AgentAppPort)
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:230]: log  http

File: cmd/agent/cmd/appHttpServer.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+	api.EchoPostHandler = &echoPostHandler{logger: logger.Named("route: summary counts")}
+	api.EchoPutHandler = &echoPutHandler{logger: logger.Named("route: summary counts")}
+	api.EchoHeadHandler = &echoHeadHandler{logger: logger.Named("route: summary counts")}
+	api.EchoOptionsHandler = &echoOptionsHandler{logger: logger.Named("route: summary counts")}
+	api.EchoPatchHandler = &echoPatchHandler{logger: logger.Named("route: summary counts")}
+
+	srvApp := server.NewServer(api)
+	srvApp.EnabledListeners = []string{types.AgentConfig.AgentAppProtocol}
+	if types.AgentConfig.AgentAppProtocol == "https" {
+		srvApp.TLSPort = int(types.AgentConfig.AgentAppPort)
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:221]: log  https

File: cmd/agent/cmd/config.go [hunk 6]
```diff
@@ -71,6 +71,10 @@ func init() {
 	globalFlag := rootCmd.PersistentFlags()
 	globalFlag.StringVarP(&types.AgentConfig.ConfigMapPath, "config-path", "C", "", "configmap file path")
 	globalFlag.BoolVarP(&types.AgentConfig.AppMode, "app-mode", "A", false, "app mode")
+	globalFlag.StringVarP(&types.AgentConfig.AgentAppProtocol, "app-protocol", "P", "http", "app mode protocol")
```
Comment (weizhoublue) [inline cmd/agent/cmd/config.go:74]: AgentAppProtocol -> enableHttpTls : true / false

File: cmd/agent/cmd/daemon.go [hunk 7]
```diff
@@ -24,18 +24,17 @@ func SetupUtility() {
...
-
 	if types.AgentConfig.AppMode {
 		// app mode, just used to debug
 		rootLogger.Info("run in app mode")
 		// sleep forever
 		select {}
 	} else {
 		rootLogger.Info("run in agent mode")
-
+		initGrpcServer()
```
Comment (weizhoublue) [inline cmd/agent/cmd/daemon.go:37]: initGrpcServer also could be test in app mode

File: cmd/agent/cmd/appHttpServer.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	p := echo.NewPutParams()
+	p.HTTPRequest = r.HTTPRequest
+	p.Delay = r.Delay
+	return (&echoPutHandler{logger: s.logger}).Handle(p)
+}
+
+func SetupAppHttpServer() {
+	logger := rootLogger.Named("app http")
+
+	if types.AgentConfig.AgentAppPort == 0 {
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:193]: AgentAppPort -> AppHttpPort

File: cmd/agent/cmd/appHttpServer.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+	api.EchoDeleteHandler = &echoDeleteHandler{logger: logger.Named("route: summary counts")}
+	api.EchoPostHandler = &echoPostHandler{logger: logger.Named("route: summary counts")}
+	api.EchoPutHandler = &echoPutHandler{logger: logger.Named("route: summary counts")}
+	api.EchoHeadHandler = &echoHeadHandler{logger: logger.Named("route: summary counts")}
+	api.EchoOptionsHandler = &echoOptionsHandler{logger: logger.Named("route: summary counts")}
+	api.EchoPatchHandler = &echoPatchHandler{logger: logger.Named("route: summary counts")}
+
+	srvApp := server.NewServer(api)
+	srvApp.EnabledListeners = []string{types.AgentConfig.AgentAppProtocol}
+	if types.AgentConfig.AgentAppProtocol == "https" {
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:220]: when tls is enabled, http 1.1 and http2.0 is supported at the same time ?

File: cmd/agent/cmd/appHttpServer.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+	atomic.AddInt64(&requestCounts, 1)
+	t := echo.NewGetOK()
+	t.Payload = &models.EchoRes{
+		ClientIP:      r.HTTPRequest.RemoteAddr,
+		RequestHeader: head,
+		RequestURL:    r.HTTPRequest.RequestURI,
+		RequestMethod: r.HTTPRequest.Method,
+		ServerName:    hostname,
+		RequestCount:  atomic.LoadInt64(&requestCounts),
+		ParamDetail:   ParamInformation,
```
[the remark this answers is not in the document]
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go:58]: CONNECT 这个方法不支持 剩下的都支持

File: cmd/agent/cmd/appHttpServer.go (hunk 9 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go:220]: 同时支持

File: charts/templates/service.yaml (hunk 1 above)
Comment (ii2day) [reply charts/templates/service.yaml]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 3 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 4 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 5 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: cmd/agent/cmd/daemon.go (hunk 7 above)
Comment (ii2day) [reply cmd/agent/cmd/daemon.go]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 8 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (weizhoublue) [reply cmd/agent/cmd/appHttpServer.go:58]: 我说 在 响应中 list 所有支持的 method

File: pkg/types/agent_config.go [hunk 11]
```diff
@@ -6,14 +6,15 @@ package types
 var AgentEnvMapping = []EnvMapping{
 	{"ENV_ENABLED_METRIC", "false", &AgentConfig.EnableMetric},
 	{"ENV_METRIC_HTTP_PORT", "", &AgentConfig.MetricPort},
-	{"ENV_HTTP_PORT", "80", &AgentConfig.HttpPort},
+	{"ENV_AGENT_HEALTH_HTTP_PORT", "80", &AgentConfig.AgentHealthPort},
```
Comment (weizhoublue) [inline pkg/types/agent_config.go:9]: health 端 不建议占用 80

4987 之类的

File: pkg/types/agent_config.go [hunk 12]
```diff
@@ -6,14 +6,15 @@ package types
...
 	{"ENV_GOPS_LISTEN_PORT", "", &AgentConfig.GopsPort},
 	{"ENV_WEBHOOK_PORT", "", &AgentConfig.WebhookPort},
 	{"ENV_PYROSCOPE_PUSH_SERVER_ADDRESS", "", &AgentConfig.PyroscopeServerAddress},
 	{"ENV_POD_NAME", "", &AgentConfig.PodName},
 	{"ENV_POD_NAMESPACE", "", &AgentConfig.PodNamespace},
 	{"ENV_GOLANG_MAXPROCS", "8", &AgentConfig.GolangMaxProcs},
 	{"ENV_AGENT_GRPC_LISTEN_PORT", "3000", &AgentConfig.AgentGrpcListenPort},
+	{"ENV_AGENT_APP_HTTP_PORT", "8080", &AgentConfig.AppHttpPort},
```
Comment (weizhoublue) [inline pkg/types/agent_config.go:17]: 如果是 http ， 占用 80
如果是 https ， 占用 443

File: charts/templates/daemonset.yaml [hunk 13]
```diff
@@ -147,8 +150,10 @@ spec:
               value: {{ .Values.kdoctorAgent.prometheus.enabled | quote }}
             - name: ENV_METRIC_HTTP_PORT
               value: {{ .Values.kdoctorAgent.prometheus.port | quote }}
-            - name: ENV_HTTP_PORT
-              value: {{ .Values.kdoctorAgent.httpServer.port | quote }}
+            - name: ENV_AGENT_HEALTH_HTTP_PORT
+              value: {{ .Values.kdoctorAgent.httpServer.healthPort | quote }}
+            - name: ENV_AGENT_APP_HTTP_PORT
```
Comment (weizhoublue) [inline charts/templates/daemonset.yaml:155]: ENV_AGENT_APP_HTTP_PORT  80
ENV_AGENT_APP_HTTP_PORT  443

File: charts/templates/daemonset.yaml [hunk 14]
```diff
@@ -103,14 +103,17 @@ spec:
             {{- end }}
             {{- if or .Values.feature.enableIPv4 .Values.feature.enableIPv6 }}
             - name: http
-              containerPort: {{ .Values.kdoctorAgent.httpServer.port }}
+              containerPort: {{ .Values.kdoctorAgent.httpServer.healthPort }}
+              protocol: TCP
+            - name: app-http
```
Comment (weizhoublue) [inline charts/templates/daemonset.yaml:108]: - name: http
- name: https

File: cmd/agent/cmd/appHttpServer.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: pkg/types/agent_config.go (hunk 11 above)
Comment (ii2day) [reply pkg/types/agent_config.go]: done

File: pkg/types/agent_config.go (hunk 12 above)
Comment (ii2day) [reply pkg/types/agent_config.go]: done

File: charts/templates/daemonset.yaml (hunk 13 above)
Comment (ii2day) [reply charts/templates/daemonset.yaml:160]: done

File: charts/templates/daemonset.yaml (hunk 14 above)
Comment (ii2day) [reply charts/templates/daemonset.yaml]: done

**Comment by ii2day** [comment]
#32

File: charts/values.yaml [hunk 15]
```diff
@@ -167,7 +167,7 @@ kdoctorAgent:
     ingressClass: ""
 
     ## @param kdoctorAgent.ingress.route the route of agent ingress. Default to "/kdoctoragent", if it changes, ingress please re-write url forwarded to "/kdoctoragent"
-    route: "/kdoctoragent"
```
Comment (weizhoublue) [inline charts/values.yaml:170]: ????

File: charts/values.yaml [hunk 16]
```diff
@@ -323,8 +327,31 @@ kdoctorAgent:
     ## @param kdoctorAgent.debug.gopsPort the gops port of template agent
     gopsPort: 5712
 
-## @section kdoctorController parameters
-##
```
Comment (weizhoublue) [inline charts/values.yaml:327]: ????

File: test/yaml/testbody.yaml [hunk 17]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+data:
+  test1: test1
+  test2: test2
+kind: ConfigMap
+metadata:
+  name: kdoctor-test-body
```
Comment (weizhoublue) [inline test/yaml/testbody.yaml:7]: ??

File: test/Makefile [hunk 18]
```diff
@@ -510,10 +510,14 @@ install_nginx:
...
+install_example_app: IMAGE_TAG ?= $(GIT_COMMIT_VERSION)
 install_example_app:
+	@echo "---------- install example app configmap request body"
+	kubectl --kubeconfig=$(KIND_KUBECONFIG) apply -f yaml/testbody.yaml
+
 	@echo "---------- install example app"
-	sed 's?<<EXAMPLE_APP_IMAGE>>?ghcr.io/kdoctor-io/kdoctor-agent:'$(PROJECT_IMAGE_VERSION)'?' yaml/testpod.yaml | kubectl --kubeconfig=$(KIND_KUBECONFIG) apply -f -
-	@ if ! kubectl rollout status  daemonset/test --kubeconfig $(KIND_KUBECONFIG) -w --timeout=120s ; then \
+	sed 's?<<EXAMPLE_APP_IMAGE>>?ghcr.io/kdoctor-io/kdoctor-agent:'$(IMAGE_TAG)'?' yaml/testpod.yaml | kubectl --kubeconfig=$(KIND_KUBECONFIG) apply -f -
+	@ if ! kubectl rollout status  daemonset/kdoctor-test -n kube-system --kubeconfig $(KIND_KUBECONFIG) -w --timeout=120s ; then \
```
Comment (weizhoublue) [inline test/Makefile:520]: 就一个 app，导致 CI 不能串行，否则 所有的压测数据都不准了

File: cmd/agent/cmd/appHttpServer.go [hunk 19]
```diff
@@ -0,0 +1,348 @@
+// Copyright 2023 Authors of kdoctor-io
+// SPDX-License-Identifier: Apache-2.0
+
+package cmd
```
Comment (weizhoublue) [inline cmd/agent/cmd/appHttpServer.go:4]: 这个实现 越来越大了，可 放到 pkg 中 ？

File: charts/values.yaml [hunk 20]
```diff
@@ -26,6 +26,21 @@ global:
   ## @param global.configName the configmap name
   configName: "kdoctor"
 
+  ## TLS configuration for kdoctor
+  tls:
```
Comment (weizhoublue) [inline charts/values.yaml:30]: tls 不是 一个子层级，是一个 顶级的层级

File: charts/values.yaml [hunk 21]
```diff
@@ -26,6 +26,21 @@ global:
...
+      secretName: "kdoctor-ca"
+
+    ## TLS configuration for kdoctor client
+    client:
+      ## @param kdoctorAgent.tls.client.secretName the secret name for storing TLS certificates
+      secretName: "kdoctor-client-cert"
+
+      ## @param kdoctorAgent.tls.client.certExpiration server cert expiration
+      # in day, default 200 years
+      certExpiration: '73000'
```
Comment (weizhoublue) [inline charts/values.yaml:42]: 不需要了，我们没有 刷 新的机制，暴露给用户是累赘

File: pkg/utils/cert.go [hunk 22]
```diff
@@ -23,46 +23,74 @@ import (
...
-func NewServerCertKey(host string, alternateIPs []net.IP, alternateDNS []string) (serverCert, serverKey, caCert []byte, err error) {
-
-	// ----------- generate self-signed ca
+func NewServerCertKey(host string, alternateIPs []net.IP, alternateDNS []string, caCertPath, caKeyPath string) (serverCert, serverKey, caCert []byte, err error) {
 	validFrom := time.Now().Add(-time.Hour) // valid an hour earlier to avoid flakes due to clock skew
+	var caCertificate *x509.Certificate
+	var caKey *rsa.PrivateKey
+	var caDERBytes []byte
+	// ----------- generate self-signed ca
+	if caCertPath == "" && caKeyPath == "" {
```
Comment (weizhoublue) [inline pkg/utils/cert.go:37]: ||

File: pkg/utils/cert.go [hunk 23]
```diff
@@ -23,46 +23,74 @@ import (
...
+
+		caKey, err := rsa.GenerateKey(cryptorand.Reader, 2048)
```
Comment (weizhoublue) [inline pkg/utils/cert.go:39]: 日志，generate self-signed ca

File: pkg/utils/cert.go [hunk 24]
```diff
@@ -23,46 +23,74 @@ import (
...
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		caCertificate, err = x509.ParseCertificate(caDERBytes)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+	} else {
+		//read ca cert
```
Comment (weizhoublue) [inline pkg/utils/cert.go:72]: 日志

File: cmd/agent/cmd/cert.go [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+		// ipv6 dns
+		alternateDNS = append(alternateDNS, types.AgentConfig.Configmap.AgentSerivceIpv6Name)
+		domain := fmt.Sprintf("%s.%s", types.AgentConfig.Configmap.AgentSerivceIpv6Name, types.AgentConfig.PodNamespace)
+		alternateDNS = append(alternateDNS, domain)
+		alternateDNS = append(alternateDNS, fmt.Sprintf("%s.svc", domain))
+		alternateDNS = append(alternateDNS, fmt.Sprintf("%s.svc.cluster.local", domain))
+	}
+
+	alternateDNS = append(alternateDNS, types.AgentConfig.PodName)
+	logger.Sugar().Debugf("alternate ip  %v", alternateIP)
```
Comment (weizhoublue) [inline cmd/agent/cmd/cert.go:98]: alternate ip for tls cert:
alternate dns for tls cert

File: charts/values.yaml (hunk 15 above)
Comment (ii2day) [reply charts/values.yaml]: done

File: charts/values.yaml (hunk 16 above)
Comment (ii2day) [reply charts/values.yaml:327]: done

File: test/yaml/testbody.yaml (hunk 17 above)
Comment (ii2day) [reply test/yaml/testbody.yaml]: done

File: cmd/agent/cmd/appHttpServer.go (hunk 19 above)
Comment (ii2day) [reply cmd/agent/cmd/appHttpServer.go]: done

File: charts/values.yaml (hunk 20 above)
Comment (ii2day) [reply charts/values.yaml]: done

File: charts/values.yaml (hunk 21 above)
Comment (ii2day) [reply charts/values.yaml]: done

File: pkg/utils/cert.go (hunk 22 above)
Comment (ii2day) [reply pkg/utils/cert.go]: done

File: cmd/agent/cmd/cert.go (hunk 25 above)
Comment (ii2day) [reply cmd/agent/cmd/cert.go]: done

File: cmd/agent/cmd/config.go [hunk 26]
```diff
@@ -71,6 +71,9 @@ func init() {
 	globalFlag := rootCmd.PersistentFlags()
 	globalFlag.StringVarP(&types.AgentConfig.ConfigMapPath, "config-path", "C", "", "configmap file path")
 	globalFlag.BoolVarP(&types.AgentConfig.AppMode, "app-mode", "A", false, "app mode")
+	globalFlag.BoolVarP(&types.AgentConfig.Insecure, "insecure", "K", true, "skip verify tls")
```
Comment (weizhoublue) [inline cmd/agent/cmd/config.go:74]: tls-insecure

**Pull request merged**